api/src/main/java/io/cloudevents/Extension.java

```diff
@@ -17,12 +17,39 @@
 
 package io.cloudevents;
 
-import java.util.Map;
+import io.cloudevents.lang.Nullable;
 
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.Set;
+
+/**
+ * Materialized CloudEvent Extension interface to read/write the extension attributes key/values.
+ */
+@ParametersAreNonnullByDefault
 public interface Extension {
 
-    void readFromEvent(CloudEvent event);
+    /**
+     * Fill this materialized extension with values from a {@link CloudEventExtensions} implementation
+     *
+     * @param extensions
+     */
+    void readFrom(CloudEventExtensions extensions);
+
+    /**
+     * Get the attribute of extension named {@code key}
+     *
+     * @param key the name of the extension attribute
+     * @return the extension value
+     * @throws IllegalArgumentException if the key is not a valid attribute key for this extension
```

Comment:
How about this:
```suggestion
     * @throws IllegalArgumentException if the key is unknown to this extension
```

to avoid additional questions on "validity" and "attribute"